gateway/src/apicast/policy/rewrite_url_captures/named_args_matcher.lua

```diff
@@ -0,0 +1,115 @@
+local re_gsub = ngx.re.gsub
+local re_match = ngx.re.match
+local re_gmatch = ngx.re.gmatch
+local re_split = require('ngx.re').split
+local insert = table.insert
+local unpack = unpack
+local ipairs = ipairs
+local setmetatable = setmetatable
+
+local _M = {}
+
+local mt = { __index = _M }
+
+-- Returns a list of named args extracted from a match_rule.
+-- For example, for the rule /{abc}/{def}?{ghi}=1, it returns this list:
+-- { "abc", "def", "ghi" }.
+local function extract_named_args(match_rule)
+  local iterator, err = re_gmatch(match_rule, [[\{(.+?)\}]], 'oj')
+
+  if not iterator then
+    return nil, err
+  end
+
+  local named_args = {}
+
+  while true do
+    local m, err_iter = iterator()
+    if err_iter then
+      return nil, err_iter
+    end
+
+    if not m then
+      break
+    end
+
+    insert(named_args, m[1])
+  end
+
+  return named_args
+end
+
+-- Rules contain {} for named args. This function replaces those with "()" to
+-- be able to capture those args when matching the regex.
+local function transform_rule_to_regex(match_rule)
+  return re_gsub(
+    match_rule,
+    [[\{.+?\}]],
+    [[([\w-.~%!$$&'()*+,;=@:]+)]], -- Same as in the MappingRule module
+    'oj'
+  )
+end
+
+-- Transforms a string representing the args of a query like:
+-- "a=1&b=2&c=3" into a table: { a = "1", b = "2", c = "3" }.
+local function string_params_to_table(string_params)
+  local res = {}
+
+  local params_split = re_split(string_params, '&')
+
+  for _, param in ipairs(params_split) do
+    local name, val = unpack(re_split(param, '='))
```

Comment:
Another unpack in a loop. This will prevent JIT.

Reply:
Fixed :+1: